### PR TITLE
Ignore version numbers in branch names that are not release branches.

### LINKED ISSUE #1536: Why do version numbers in branch names change the calculated version?
Problem
---------

Since GitVersion 4 we noticed a suddend jump in the version generated by GitVersion. Looking at the diagnostic output it seems to respect a version number in the name of a merged branch. That is not intended though, because the version number in the branch name is purely coincidental:
`bugfix/COCO-1266-ng-bootstrap-ng-bootstrap-2.2.0`.

The `2.2.0` in the branch name seems to cause a bump to `2.3.0` on `develop`:

```
> gitversion /showvariable FullSemVer
2.3.0-alpha.1351
```

The previous version of GitVersion correctly calculated `1.8.0`.

Question
----------

I'm wondering whether this behavior can be configured? Maybe there is a regex somewhere that we need to adjust?

Additional Info
----------------

I figured we could set `next-version` or use a git tag as a workaround.

We are currently using an "unconfigured" (no `GitVersion.yml`) installation of GitVersion 4.0.0:

```
> choco list -l GitVersion.Portable
Chocolatey v0.10.8
GitVersion.Portable 4.0.0
1 packages installed.
```

Here's the diagnostic output:

```
> gitversion /showvariable FullSemVer /diag
INFO [11/19/18 13:58:01:60] Dumping commit graph: 
...

INFO [11/19/18 13:58:01:72] Working directory: C:\dev\contracting
INFO [11/19/18 13:58:01:72] IsDynamicGitRepository: False
INFO [11/19/18 13:58:01:74] Returning Project Root from DotGitDirectory: C:\dev\contracting\.git - C:\dev\contracting
INFO [11/19/18 13:58:01:74] Running on Windows.
INFO [11/19/18 13:58:01:74] IsDynamicGitRepository: False
INFO [11/19/18 13:58:01:74] Returning Project Root from DotGitDirectory: C:\dev\contracting\.git - C:\dev\contracting
INFO [11/19/18 13:58:01:74] Project root is: C:\dev\contracting
INFO [11/19/18 13:58:01:74] DotGit directory is: C:\dev\contracting\.git
INFO [11/19/18 13:58:01:75] IsDynamicGitRepository: False
INFO [11/19/18 13:58:01:75] Returning Project Root from DotGitDirectory: C:\dev\contracting\.git - C:\dev\contracting
INFO [11/19/18 13:58:01:79] IsDynamicGitRepository: False
INFO [11/19/18 13:58:01:79] Returning Project Root from DotGitDirectory: C:\dev\contracting\.git - C:\dev\contracting
INFO [11/19/18 13:58:01:80] Using latest commit on specified branch
INFO [11/19/18 13:58:01:82] Running against branch: develop (a812a30ef3fe2fa6e8c98d11edac766a977ebc9f)
INFO [11/19/18 13:58:01:83] Begin: Calculating base versions
  INFO [11/19/18 13:58:01:97] Fallback base version: 0.1.0 with commit count source 273b1c079ff02f564f54b3cb06c272f07116d8f4 (Incremented: None)
  INFO [11/19/18 13:58:02:31] Git tag '1.6.0': 1.6.0 with commit count source 4af704417ba5a3d9dce22599d32f0db5f0018bdf (Incremented: 1.7.0)
  INFO [11/19/18 13:58:02:33] Git tag '1.5.0': 1.5.0 with commit count source b5f6f313cea96adbcd3c7a541fdb9728fb1e51da (Incremented: 1.6.0)
  INFO [11/19/18 13:58:02:36] Git tag '1.4.0': 1.4.0 with commit count source c391c7e7b9e76cfd7b0478e06b4aa7713abc7687 (Incremented: 1.5.0)
  INFO [11/19/18 13:58:02:39] Git tag '1.3.0': 1.3.0 with commit count source 3b7bd6a70184061f1780752017fcaed8fe0c7ae8 (Incremented: 1.4.0)
  INFO [11/19/18 13:58:02:56] Merge message 'Merge pull request #159 in ITD/contracting from bugfix/COCO-1266-ng-bootstrap-ng-bootstrap-2.2.0 to develop

* commit '079e3ed489ce4da8195e20f9225030155128eefd':
  COCO-1266: changed css so card-header will have less height again.': 2.2.0 with commit count source 795ef6078ed0a7ccba1ff9f473a80cd884cf3e54 (Incremented: 2.3.0)
  INFO [11/19/18 13:58:02:66] Git tag '1.7.0': 1.7.0 with commit count source d59beaf2be8ae73bc6e7a21e30c66e87d93cf645 (Incremented: 1.8.0)
  INFO [11/19/18 13:58:02:67] Git tag '1.6.0': 1.6.0 with commit count source 4af704417ba5a3d9dce22599d32f0db5f0018bdf (Incremented: 1.7.0)
  INFO [11/19/18 13:58:02:69] Git tag '1.5.0': 1.5.0 with commit count source b5f6f313cea96adbcd3c7a541fdb9728fb1e51da (Incremented: 1.6.0)
  INFO [11/19/18 13:58:02:72] Git tag '1.4.0': 1.4.0 with commit count source c391c7e7b9e76cfd7b0478e06b4aa7713abc7687 (Incremented: 1.5.0)
  INFO [11/19/18 13:58:02:75] Git tag '1.3.0': 1.3.0 with commit count source 3b7bd6a70184061f1780752017fcaed8fe0c7ae8 (Incremented: 1.4.0)
  INFO [11/19/18 13:58:02:77] Found multiple base versions which will produce the same SemVer (2.3.0), taking oldest source for commit counting (Merge message 'Merge pull request #159 in ITD/contracting from bugfix/COCO-1266-ng-bootstrap-ng-bootstrap-2.2.0 to develop

* commit '079e3ed489ce4da8195e20f9225030155128eefd':
  COCO-1266: changed css so card-header will have less height again.')
  INFO [11/19/18 13:58:02:78] Base version used: Merge message 'Merge pull request #159 in ITD/contracting from bugfix/COCO-1266-ng-bootstrap-ng-bootstrap-2.2.0 to develop

* commit '079e3ed489ce4da8195e20f9225030155128eefd':
  COCO-1266: changed css so card-header will have less height again.': 2.2.0 with commit count source 795ef6078ed0a7ccba1ff9f473a80cd884cf3e54 (Incremented: 2.3.0)
INFO [11/19/18 13:58:02:78] End: Calculating base versions (Took: 945.00ms)
INFO [11/19/18 13:58:02:83] 1351 commits found between 795ef6078ed0a7ccba1ff9f473a80cd884cf3e54 and a812a30ef3fe2fa6e8c98d11edac766a977ebc9f
INFO [11/19/18 13:58:02:84] Begin: Getting version tags from branch 'refs/heads/develop'.
INFO [11/19/18 13:58:02:94] End: Getting version tags from branch 'refs/heads/develop'. (Took: 100.00ms)
```

### PR DESCRIPTION
**Pull request by bert2**
This fixes #1536.

~~Introduces a new config switch in the [`ignore` section](https://github.com/bert2/GitVersion/blob/bugfix/issue-1536-ignore-versions-in-branch-names/docs/configuration.md#ignore) called [`non-release-branches`](https://github.com/bert2/GitVersion/blob/bugfix/issue-1536-ignore-versions-in-branch-names/docs/configuration.md#non-release-branches). Setting it to `true` will ignore versions coming from branch names that are not [release branches](https://github.com/bert2/GitVersion/blob/bugfix/issue-1536-ignore-versions-in-branch-names/docs/configuration.md#is-release-branch).~~

Branches that are not [release branches](https://github.com/bert2/GitVersion/blob/bugfix/issue-1536-ignore-versions-in-branch-names/docs/configuration.md#is-release-branch) will not be considered for version calculation anymore.

Works with versions coming from branch names in merge commits and from the currently checked out branch.

Also works for remote-tracking branches as long as their remote is `origin`.

**Additional notes:**

* ~~There is some duplication in the strategies `MergeMessageBaseVersionStrategy` and `VersionInBranchNameBaseVersionStrategy` that I didn't know where to extract it to.~~
    * The `Config` class seemed like a good place.
* ~~The check if a branch is a release branch fails when a remote branch is merged due to the `origin/` prefix in the merge commit message: `"Merge remote-tracking branch 'origin/release/0.8.0' into develop/master"`. Should such a merge be considered a merge of a release branch? Please let me know and I'll ammend the PR.~~
    * Fixed by removing `"refs/remotes/<remote-name>/"` from the branch name if present. For instance `"refs/remotes/origin/release/2.0.0"` will be stripped to it's friendly name without the remote, i.e. `"release/2.0.0"`.
* ~~I'm not sure how some branches should be handled. Originally the branch `Particular/release-1.0.0` would always yield the version `1.0.0`. With this PR the above branch name will only be considered for version calculation when `Particular` also is a remote reference of the repository. Is that assumption correct?~~
    * Only merges of remote-tracking release branches coming from `origin` can yield versions. All other remotes are ignored.

### CODE REVIEW

**Comment by bert2** [comment]
Something ain't right with Travis... He failed the first build with a random error in the linux job while the osx job went through.
And now he just failed the osx job with another random error, but the linux job passed?

**Comment by asbjornu** [comment]
@bert2: I ran the build again, so it's not successful. If you can `rebase` against the latest `master`, this LGTM. Thanks! 😃

**Comment by asbjornu** [comment]
As #1422 is close to being merged and we discuss incrementing major with it, I think we should reverse the current behavior and ignore the version number in branches that aren't release branches by default. I'm not even sure we need a config for enabling the current (imho unintuitive) behavior. Thoughts, @bert2?

**Comment by asbjornu** [comment]
@bert2: Yes, whether the branch is remote or local shouldn't matter in its consideration for release classification, I think. So remote or local are just the same; a release branch is a release branch no matter where it lives.

Just take this up when you have time for it and when you do, I suggest you rebase on `master` (I hope #1422 will be merged by then), remove the config and reverse the logic so it applies by default.

File: src/GitVersionCore.Tests/Mocks/MockRepository.cs [hunk 1]
```diff
@@ -1,5 +1,8 @@
 using System;
 using System.Collections.Generic;
+
+using GitVersionCore.Tests.Mocks;
```
Comment (bert2) [inline src/GitVersionCore.Tests/Mocks/MockRepository.cs]: The `Mock*` classes are all without namespace. I accidentally created the new ones with the sub-namespace `Mock`, but will fix this later.

**Comment by bert2** [comment]
Hey @asbjornu, I changed the PR as suggested and updated the description above. Three tests are still failing, because I'm not sure what the expected behavior should be. Please see [this comment](https://github.com/GitTools/GitVersion/pull/1541/files/9fd376c8110d7ab3e87b310bf781be9f4f0e3532#r243334099).

File: src/GitVersionCore.Tests/VersionCalculation/Strategies/MergeMessageBaseVersionStrategyTests.cs [hunk 2]
```diff
@@ -30,38 +32,25 @@ public void ShouldNotAllowIncrementOfVersion()
...
-        [TestCase("Merge pull request #95 in Particular/issue-94", false, null)]
-        [TestCase("Merge pull request #64 from arledesma/feature-VS2013_3rd_party_test_framework_support", true, null)]
         [TestCase("Merge pull request #500 in FOO/bar from Particular/release-1.0.0 to develop)", true, "1.0.0")]
-        [TestCase("Merge pull request #500 in FOO/bar from feature/new-service to develop)", true, null)]
         [TestCase("Finish Release-0.12.0", true, "0.12.0")] //Support Syntevo SmartGit/Hg's Gitflow merge commit messages for finishing a 'Release' branch
-        [TestCase("Finish 0.14.1", true, "0.14.1")] //Support Syntevo SmartGit/Hg's Gitflow merge commit messages for finishing a 'Hotfix' branch
         [TestCase("Merge branch 'Release-v0.2.0'", true, "0.2.0")]
         [TestCase("Merge branch 'Release-v2.2'", true, "2.2.0")]
         [TestCase("Merge remote-tracking branch 'origin/release/0.8.0' into develop/master", true, "0.8.0")]
-        public void AssertMergeMessage(string message, bool isMergeCommit, string expectedVersion)
```
[the remark this answers is not in the document]
Comment (asbjornu) [reply src/GitVersionCore.Tests/VersionCalculation/Strategies/MergeMessageBaseVersionStrategyTests.cs]: Yes, I agree; `Particular` should be considered a remote ref that should not affect the version number by default. This is a breaking change we're introducing here and I think it's fine.

[the remark this answers is not in the document]
Comment (bert2) [reply src/GitVersionCore.Tests/VersionCalculation/Strategies/MergeMessageBaseVersionStrategyTests.cs]: @asbjornu, I'm not sure I understand. Currently _all_ remote refs of a repo are considered when shortening a branch name to its friendly name.

So when `Particular` is listed as `[remote "Particular"]` in the git config of the repo then the branch name `Particular/release-1.0.0` will be reduced to `release-1.0.0` and thus be considered a release branch affecting the version.

This is done by the new method [`AsBranchNameWithoutRemote()`](https://github.com/GitTools/GitVersion/pull/1541/files/9fd376c8110d7ab3e87b310bf781be9f4f0e3532#diff-82df1f328d9f7f5a1f9b55d7d88d7e38).

If I understand you correctly then only "special" remotes like `origin` and maybe `upstream` should be allowed to affect the version. In that case the above methods needs to be changed.

[the remark this answers is not in the document]
Comment (bert2) [reply src/GitVersionCore.Tests/VersionCalculation/Strategies/MergeMessageBaseVersionStrategyTests.cs]: Ok, nice. This makes things a little easier then.

**Comment by asbjornu** [comment]
Excellent. We now just need to figure out #1550 and we should be good to go. Thanks for your patience and all the work you’ve put into this! :pray:

**Comment by richardgavel** [comment]
Any chance this will be able to be merged soon?

**Comment by asbjornu** [comment]
Now that #1581 is merged, I think we're in a better place wrt. #1550. What do you say, @arturcic? Do we have a working product again? I'm sorry I don't have time to test this myself at the moment. 😞

**Comment by arturcic** [comment]
More or less yes, we still use older version of Libgit2sharp, and I have integrated the code from GitTools.Core to use directly the Libgit2Sharp library. I'm planning to upgrade it and fix the packaging. For now the current version relies on the workaround to explicitly install the needed libgit2-library on linux

**Comment by arturcic** [comment]
I guess we can integrate some PRs that improve the code, as the changes I'm planning are more related to build/packaging

**Comment by asbjornu** [comment]
@arturcic, Ok, sounds good. The only thing missing from #1422 then is https://github.com/GitTools/GitVersion/pull/1422#issuecomment-449349025. I guess any commit is as good as any for a `5.0.0-alpha.1` tag?

**Comment by arturcic** [comment]
> @arturcic, Ok, sounds good. The only thing missing from #1422 then is [#1422 (comment)](https://github.com/GitTools/GitVersion/pull/1422#issuecomment-449349025). I guess any commit is as good as any for a `5.0.0-alpha.1` tag?

Agree

**Comment by asbjornu** [comment]
Since #1422 marked the beginning of this whole debacle, I tagged cf2ccfa2db2f4628259fc170cfba8de0f040c551 with `5.0.0-alpha.1`.

**Pull request merged**